File: src/main/scala/org/winlogon/powertools/PowerToolsPlugin.scala
package org.winlogon.powertools

import dev.jorel.commandapi.CommandAPICommand
import dev.jorel.commandapi.arguments.{
    EnchantmentArgument, GreedyStringArgument,
    IntegerArgument, PlayerArgument, StringArgument,
    ArgumentSuggestions
}
import dev.jorel.commandapi.executors.{CommandArguments, CommandExecutor}

import org.bukkit.attribute.Attribute
import org.bukkit.command.{CommandSender, ConsoleCommandSender}
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.EnchantmentStorageMeta
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.{Bukkit, Material}

import io.papermc.paper.chat.ChatRenderer
import io.papermc.paper.datacomponent.DataComponentTypes
import io.papermc.paper.event.player.AsyncChatEvent

import net.kyori.adventure.audience.Audience
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer

import scala.jdk.CollectionConverters.*
import scala.util.{Try, Success, Failure}
import scala.util.boundary
import scala.util.chaining.scalaUtilChainingOps
import org.bukkit.event.inventory.InventoryType

case class UnenchantConfig(basePrice: Double)
case class UnsafeEnchantConfig(enabled: Boolean)
case class HealConfig(removeEffects: Boolean, showWhoHealed: Boolean)
case class TransferConfig(enabled: Boolean, servers: List[java.util.HashMap[String, String]])
case class Configuration(
    heal: HealConfig,
    unenchant: UnenchantConfig,
    unsafeEnchants: UnsafeEnchantConfig,
    transferConfig: TransferConfig
)

class PowerToolsPlugin extends JavaPlugin {
    private var config: Configuration = _

    override def onEnable(): Unit = {
        config = loadConfig()
        registerCommands()
    }

    private def loadConfig(): Configuration = {
        saveDefaultConfig()
        reloadConfig()
        val yamlConfig = getConfig

        val healCfg = HealConfig(
            removeEffects = yamlConfig.getBoolean("heal.remove-effects", true),
            showWhoHealed = yamlConfig.getBoolean("heal.show-who-healed", false)
        )
        val unenchantCfg = UnenchantConfig(yamlConfig.getDouble("unenchant.base-price", 5.0))
        val unsafeCfg = UnsafeEnchantConfig(yamlConfig.getBoolean("unsafe-enchants.enabled", true))
        val transferCfg = TransferConfig(
            enabled = yamlConfig.getBoolean("transfer.enabled", true),
            servers = yamlConfig.getMapList("transfer.servers").asScala.map { configMap =>
                val map = new java.util.HashMap[String, String]()
                configMap.asScala.foreach { case (k, v) =>
                    map.put(k.toString, v.toString)
                }
                map
            }.toList
        )

        Configuration(healCfg, unenchantCfg, unsafeCfg, transferCfg)
    }

    // TODO: add user-facing /transfer command which is configurable
    private def registerCommands(): Unit = {
        // Helper to always return 1 as status.
        val successStatus: Int = 1

        // Broadcast Command
        CommandAPICommand("broadcast")
            .withAliases("bc")
            .withPermission("powertools.broadcast")
            .withArguments(GreedyStringArgument("message"))
            .executes((sender: CommandSender, args: CommandArguments) => {
                val message = args.get("message").asInstanceOf[String]
                executeBroadcast(sender, message)
                successStatus
            })
            .register()

        // Hat command
        CommandAPICommand("hat")
            .withPermission("powertools.hat")
            .executesPlayer((player: Player, _: CommandArguments) => {
                executeHat(player)
                successStatus
            })
            .register()

        // Invsee Command
        CommandAPICommand("invsee")
            .withPermission("powertools.invsee")
            .withArguments(PlayerArgument("target"))
            .executesPlayer((player: Player, args: CommandArguments) => {
                val targetName = Option(args.get("target").asInstanceOf[Player]).map(_.getName)
                    .getOrElse("unknown")

                if (!targetName.equalsIgnoreCase(player.getName)) {
                    executeInvsee(player, targetName)
                } else {
                    sendError(player, "you cannot invsee yourself")
                }

                successStatus
            })
            .register()

        CommandAPICommand("absorb")
            .withPermission("powertools.absorb")
            .executesPlayer((player: Player, _: CommandArguments) => {
                // TODO: absorb player and check if they have required item in inv
                successStatus
            })
            .register()

        // Smite Command
        CommandAPICommand("smite")
            .withPermission("powertools.smite")
            .withArguments(PlayerArgument("target"))
            .executes((sender: CommandSender, args: CommandArguments) => {
                val player = args.get("target").asInstanceOf[Player]
                val targetName = Option(player).map(_.getName).getOrElse("unknown")
                executeSmite(sender, targetName)
                successStatus
            })
            .register()

        // Sudo command
        CommandAPICommand("sudo")
            .withAliases("doas")
            .withPermission("powertools.wheel")
            .withArguments(StringArgument("target"))
            .withSubcommand(
                CommandAPICommand("command")
                    .withAliases("cmd")
                    .withArguments(PlayerArgument("target"))
                    .withArguments(GreedyStringArgument("command"))
                    .executes((sender: CommandSender, args: CommandArguments) => {
                        val target = args.get("target").asInstanceOf[Player]
                        val command = args.get("command").asInstanceOf[String]
                        executeSudoCommand(sender, target, command)
                        successStatus
                    })
            )
            .withSubcommand(
                CommandAPICommand("chat")
                    .withArguments(PlayerArgument("target"))
                    .withArguments(GreedyStringArgument("message"))
                    .executes((sender: CommandSender, args: CommandArguments) => {
                        val target = args.get("target").asInstanceOf[Player]
                        val message = args.get("message").asInstanceOf[String]
                        executeSudoChat(sender, target, message)
                        successStatus
                    })
            )
            .register()

        CommandAPICommand("split")
            .withPermission("powertools.splitenchants")
            .withAliases("unenchant")
            .executesPlayer((player: Player, _: CommandArguments) => {
                executeSplitUnenchant(player)
                successStatus
            })
            .register()

        CommandAPICommand("fly")
            .withPermission("powertools.fly")
            .executesPlayer((player: Player, _: CommandArguments) => {
                val canFly = player.getAllowFlight()
                val toggledFly = !canFly

                val statusMessage = if (toggledFly) "enabled" else "disabled"
                val color = if (toggledFly) "dark_aqua" else "red"

                val playerName = s"<dark_green>${player.getName()}</dark_green>"
                player.setAllowFlight(toggledFly)
                player.sendRichMessage(s"<gray>Fly <$color>$statusMessage</$color> for $playerName.</gray>")
                successStatus
            })
            .register()

        // Unsafe enchant command
        CommandAPICommand("enchantunsafe")
            .withPermission("powertools.unsafe-enchants")
            .withAliases("ue", "uenchant")
            .withArguments(EnchantmentArgument("enchantment"), IntegerArgument("level"))
            .executesPlayer((player: Player, args: CommandArguments) => {
                executeUnsafeEnchant(player, args)
                successStatus
            })
            .register()

        // Heal command
        CommandAPICommand("heal")
            .withPermission("heal.admin")
            .withAliases("h")
            .withArguments(PlayerArgument("player").setOptional(true))
            .executes((sender: CommandSender, args: CommandArguments) => {
                val maybeTarget = Option(args.get("player").asInstanceOf[Player])
                    .orElse(Option(sender).collect { case p: Player => p })
                maybeTarget.foreach(target => healPlayer(target, sender))
                successStatus
            })
            .register()

        registerTransferCommandsIfEnabled(config.transferConfig.enabled)
    }
    private def executeUnsafeEnchant(player: Player, args: CommandArguments): Unit = {
        if (!config.unsafeEnchants.enabled) {
            sendError(player, "unsafe enchantments are disabled in config")
            return
        }
        val enchantment = args.get("enchantment").asInstanceOf[Enchantment]
        val level = args.get("level").asInstanceOf[Integer].intValue()

        Option(player.getInventory.getItemInMainHand)
            .filter(item => item.getType != Material.AIR)
            .fold(sendError(player, "you must be holding an item to enchant")) { item =>
                item.addUnsafeEnchantment(enchantment, level)
                player.updateInventory()

                val displayName = PlainTextComponentSerializer.plainText()
                    .serialize(enchantment.displayName(1))

                val romanNumeralRegex = """^(?=.)M{0,4}(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})$""".r

                val finalDisplayName = {
                    val trimmed = displayName.trim
                    val words = trimmed.split("\\s+")
                    if (words.nonEmpty && romanNumeralRegex.pattern.matcher(words.last).matches()) {
                          words.dropRight(1).mkString(" ")
                    } else {
                        displayName
                    }
                }

                val prefix = "&8[&5UE&8]"
                player.sendMessage(fmt(
                    s"$prefix <gray>Applied <dark_aqua>$finalDisplayName</dark_aqua> at level <dark_green>$level</dark_green>."
                ))
            }
    }

    private def executeSplitUnenchant(player: Player): Unit = {
        val inventory = player.getInventory
        val itemHand = Option(inventory.getItemInMainHand)

        itemHand
            .filter(_.getType == Material.ENCHANTED_BOOK)
            .foreach { _ =>
                sendError(player, "the item can't be an enchanted book")
                return
            }

        itemHand
            .filterNot(item => item == null || item.getType == Material.AIR)
            .fold(sendError(player, "you must be holding an item")) { itemInHand =>
                val enchantments = itemInHand.getEnchantments
                if (enchantments.isEmpty) {
                    sendError(player, "this item has no enchantments to split")
                    return
                }
                val cost = (config.unenchant.basePrice * enchantments.size).toInt
                if (player.getTotalExperience < cost) {
                    sendError(player, s"you need at least $cost XP to split these enchantments")
                    return
                }
                player.giveExp(-cost)

                // Remove all enchantments from the item meta.
                val meta = itemInHand.getItemMeta
                enchantments.keySet.asScala.foreach(meta.removeEnchant)
                itemInHand.setItemMeta(meta)
                itemInHand.resetData(DataComponentTypes.REPAIR_COST)

                // Create an enchanted book for each enchantment.
                enchantments.asScala.foreach { case (ench, level) =>
                    val book = new ItemStack(Material.ENCHANTED_BOOK)
                    val bookMeta = book.getItemMeta.asInstanceOf[EnchantmentStorageMeta]
                    bookMeta.addStoredEnchant(ench, level, true)
                    book.setItemMeta(bookMeta)
                    // If the inventory is full, drop it at the player's position
                    if (inventory.firstEmpty() == -1)
                        player.getWorld.dropItemNaturally(player.getLocation, book)
                    else
                        inventory.addItem(book)
                }
                player.updateInventory()
                player.sendMessage(fmt(s"&7Successfully split &3${enchantments.size} &7enchantment(s) for &2$cost XP."))
            }
    }

    private def executeBroadcast(sender: CommandSender, message: String): Unit = {
        val formattedMessage = "<dark_gray>[<dark_aqua>Broadcast<dark_gray>] <message>"
        val messageComponent = Placeholder.component("message", Component.text(message, NamedTextColor.GRAY))

        Bukkit.getOnlinePlayers.forEach(_.sendRichMessage(formattedMessage, messageComponent))
        // Send message to console if the command was sent from console for user feedback
        if (sender.isInstanceOf[ConsoleCommandSender]) {
            sender.sendRichMessage(formattedMessage, messageComponent)
        }
    }

    private def executeSudoCommand(sender: CommandSender, target: Player, command: String): Unit = {
        if (target == null || !target.isOnline) {
            sendError(sender, "player not found or offline")
            return
        }
        // Use the scheduler to perform the command asynchronously.
        target.getScheduler.execute(this, () => target.chat(s"/${command.trim()}"), null, 0L)
    }

    private def executeSudoChat(sender: CommandSender, target: Player, message: String): Unit = {
        if (target == null || !target.isOnline) {
            sendError(sender, "player not found or offline")
            return
        }
        val players = Bukkit.getOnlinePlayers()

        val consoleSenderAudience = Audience.audience(Bukkit.getConsoleSender())
        val viewers = new java.util.HashSet[Audience](players)
        viewers.add(consoleSenderAudience)

        val userMsg = fmt(message)

        // `AsyncChatEvent`s called by plugins must be synchronous
        val event = AsyncChatEvent(false, target, viewers, ChatRenderer.defaultRenderer(), userMsg, userMsg, null)

        // If the event is not cancelled, send the message
        if (event.callEvent() && !event.isCancelled) {
            val renderer = event.renderer()
            val senderComponent = event.getPlayer.displayName()
            // Create an audience from the players and craft a new rendered message
            val renderedMessage = renderer.render(event.getPlayer, senderComponent, event.message(), Audience.audience(players))
            event.viewers().forEach(_.sendMessage(renderedMessage))
        }
    }

    private def executeHat(player: Player): Unit = {
        val inv = player.getInventory
        val hand = inv.getItemInMainHand
        val helmet = Option(inv.getHelmet)
        helmet match {
            case Some(h) =>
                inv.setHelmet(hand)
                inv.setItemInMainHand(h)
                player.sendRichMessage("<gray>Swapping items...</gray>")
            case None =>
                inv.setHelmet(hand)
                inv.setItemInMainHand(null)
        }
        player.updateInventory()
        player.sendRichMessage("<gray>Your held item is now <dark_aqua>on your head</dark_aqua>!")
    }

    private def executeInvsee(player: Player, targetName: String): Unit = {
        Option(Bukkit.getPlayer(targetName))
            .filter(_.isOnline)
            .fold(sendError(player, "player not found or offline"))(target => {
                var inventorySize = 45
                val targetInventory = target.getInventory

                if (targetInventory.getSize() > inventorySize) {
                    inventorySize = targetInventory.getSize()
                }

                // FIXME: you originally could open a player's inventory normally but for some
                // reason, between 1.21.5 and 1.21.6 you can't do it anymore.
                val wrapperInventory = Bukkit.createInventory(null, inventorySize)
                wrapperInventory.setContents(targetInventory.getContents)
                player.openInventory(wrapperInventory)

                // TODO: since you can't open a player inventory as before, the changes need
                // to be written back to the target's inventory
            })
    }

    private def executeSmite(sender: CommandSender, targetName: String): Unit = {
        Option(Bukkit.getPlayer(targetName))
            .filter(_.isOnline)
            .fold(sendError(sender, "player not found or offline")) { target =>
                val result = Try {
                    target.getWorld.strikeLightning(target.getLocation)
                }

                result match {
                    case Success(_) =>
                        sender.sendMessage(fmt(s"&7You have smitten &3${target.getName}!"))
                        target.sendMessage(fmt("&7You have been smitten by <b>&3a mighty force&7!"))
                    case Failure(_) =>
                        sendError(sender, "failed to smite player")
                }
            }
    }

    private def healPlayer(player: Player, sender: CommandSender): Unit = {
        Option(player.getAttribute(Attribute.MAX_HEALTH))
            .foreach(max => player.setHealth(max.getValue()))
        player.setFoodLevel(20)
        player.setSaturation(20f)

        if (config.heal.removeEffects) {
            player.getActivePotionEffects.forEach(effect => player.removePotionEffect(effect.getType))
        }

        val healedMessage = "<gray>You have been <dark_aqua>healed</dark_aqua>.</gray>"

        if (player != sender) {
            sender.sendRichMessage(
              "<gray><player> has been healed.</gray>",
              Placeholder.component("player", Component.text(player.getName(), NamedTextColor.DARK_AQUA))
            )

            val senderComponent = Component.text(sender.getName(), NamedTextColor.DARK_AQUA)

            val healMsg = if (config.heal.showWhoHealed) {
                s"<gray>You have been healed by <staff>.</gray>"
            } else {
                healedMessage
            }

            player.sendRichMessage(healMsg, Placeholder.component("staff", senderComponent))
        } else {
          sender.sendRichMessage(healedMessage)
        }
    }

    // TODO: I'm not totally sure I remember what this is here for, is it for aliasing a
    // server IP? If yes then it should have another name.
    private def registerTransferCommandsIfEnabled(transferConfigEnabled: Boolean): Unit = {
        if (!transferConfigEnabled) {
            return
        }

        CommandAPICommand("transfer")
            .withPermission("powertools.transfer")
            .withArguments(new StringArgument("server").replaceSuggestions(ArgumentSuggestions.strings(info => {
                config.transferConfig.servers.filter { server =>
                    val allowPlayers = server.getOrDefault("allowPlayers", "false").toBoolean
                    allowPlayers || info.sender().hasPermission("powertools.transfer.all")
                }.map(_.get("name")).toArray
            })))
            .executesPlayer((player: Player, args: CommandArguments) => {
                val serverName = args.get("server").asInstanceOf[String]

                config.transferConfig.servers
                    .find(_.get("name") == serverName)
                    .fold(sendError(player, "server not found"))( server => {
                        val allowPlayers = server.getOrDefault("allowPlayers", "false").toBoolean

                        if (!(allowPlayers && player.hasPermission("powertools.transfer.all"))) {
                            sendError(player, "you don't have permission to transfer to this server")
                            return
                        }

                        try {
                            val host = server.get("host")
                            val port = server.get("port").toInt
                            player.transfer(host, port)
                        } catch {
                            case _: NumberFormatException =>
                                sendError(player, "invalid port for server! Please contact a server admin")
                            case ex: Exception =>
                                sendError(player, s"transfer failed: ${ex.getMessage}")
                        }
                    })
            })
            .register()
    }

    /** Format a string converting legacy colors to MiniMessage */
    private def fmt(s: String): Component = ChatFormatting.translateLegacyCodes(s)

    /** Send an error message to the player */
    private def sendError(target: Player | CommandSender, err: String): Unit =
        target.sendMessage(fmt(s"<#F93822>Error&7: ${sentenceCase(err)}"))

    private def sentenceCase(input: String): String = {
        input
            .trim
            .pipe { s =>
              if (s.isEmpty) s
              else s.head.toUpper + s.tail
            }
            .pipe { s =>
              if (s.endsWith(".")) s else s + "."
            }
    }
}

File: src/main/kotlin/org/winlogon/powertools/PowerToolsPlugin.kt
package org.winlogon.powertools

import revxrsal.commands.Lamp
import revxrsal.commands.annotation.Command
import revxrsal.commands.annotation.Default
import revxrsal.commands.annotation.Dependency
import revxrsal.commands.annotation.Named
import revxrsal.commands.annotation.Optional
import revxrsal.commands.annotation.Subcommand
import revxrsal.commands.annotation.SuggestWith

import revxrsal.commands.bukkit.BukkitLamp
import revxrsal.commands.bukkit.actor.BukkitCommandActor
import revxrsal.commands.bukkit.annotation.CommandPermission

import net.kyori.adventure.audience.Audience
import net.kyori.adventure.key.Key
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer

import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.attribute.Attribute
import org.bukkit.command.CommandSender
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.EnchantmentStorageMeta
import org.bukkit.plugin.java.JavaPlugin
import org.winlogon.powertools.suggestions.EnchantmentSuggestions
import org.winlogon.asynccraftr.AsyncCraftr

import io.papermc.paper.datacomponent.DataComponentTypes
import io.papermc.paper.registry.RegistryAccess
import io.papermc.paper.registry.RegistryKey
import io.papermc.paper.registry.TypedKey

import kotlin.math.roundToInt

class PowerToolsPlugin : JavaPlugin() {
    private lateinit var config: Configuration
    private lateinit var absorbAnimal: AbsorbAnimal
    private lateinit var lamp: Lamp<BukkitCommandActor>
    private val romanNumeralRegex = """^(?=.)M{0,4}(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})$""".toRegex()

    override fun onEnable() {
        config = loadConfig()
        absorbAnimal = AbsorbAnimal(this)

        val sudoCommands = SudoCommands(this)
        lamp = BukkitLamp.builder(this).build()
        lamp.apply {
            register(this@PowerToolsPlugin)
            register(sudoCommands)
        }
    }

    private fun loadConfig(): Configuration {
        saveDefaultConfig()
        reloadConfig()
        val yamlConfig = getConfig()

        val healCfg = HealConfig(
            removeEffects = yamlConfig.getBoolean("heal.remove-effects", true),
            showWhoHealed = yamlConfig.getBoolean("heal.show-who-healed", false)
        )
        val unenchantCfg = UnenchantConfig(yamlConfig.getDouble("unenchant.base-price", 5.0))
        val unsafeCfg = UnsafeEnchantConfig(yamlConfig.getBoolean("unsafe-enchants.enabled", true))
        val transferCfg = TransferConfig(
            enabled = yamlConfig.getBoolean("transfer.enabled", true),
            servers = yamlConfig.getMapList("transfer.servers").map { configMap ->
                val map = HashMap<String, String>()
                configMap.forEach { (k, v) -> map[k.toString()] = v.toString() }
                map
            }
        )

        return Configuration(healCfg, unenchantCfg, unsafeCfg, transferCfg)
    }

    @Command("broadcast", "bc")
    fun broadcast(
        actor: BukkitCommandActor,
        @Named("message") message: String
    ) {
        val formattedMessage = "<dark_gray>[<dark_aqua>Broadcast<dark_gray>] <message>"
        val messageComponent = Placeholder.component("message", Component.text(message, NamedTextColor.GRAY))

        Bukkit.getOnlinePlayers().forEach { it.sendRichMessage(formattedMessage, messageComponent) }
        Bukkit.getConsoleSender().sendRichMessage(formattedMessage, messageComponent)
    }

    @Command("hat")
    fun hat(actor: BukkitCommandActor) {
        val player = actor.sender() as Player

        val inv = player.inventory
        val hand = inv.itemInMainHand
        val helmet = inv.helmet

        if (helmet != null) {
            inv.helmet = hand
            inv.setItemInMainHand(helmet)
            player.sendRichMessage("<gray>Swapping items...</gray>")
        } else {
            inv.helmet = hand
            inv.setItemInMainHand(null)
        }

        player.updateInventory()
        player.sendRichMessage("<gray>Your held item is now <dark_aqua>on your head</dark_aqua>!")
    }

    @Command("invsee")
    fun invsee(
        actor: BukkitCommandActor,
        @Named("target") target: Player
    ) {
        if (target.name == actor.name()) {
            ChatFormatting.sendError(actor.sender(), "you cannot invsee yourself")
            return
        }

        val player = actor.sender() as Player

        if (!target.isOnline) {
            ChatFormatting.sendError(player, "player not found or offline")
            return
        }

        val targetInv = target.inventory
        val inventorySize = maxOf(45, targetInv.size)

        val wrapper = Bukkit.createInventory(null, inventorySize)
        wrapper.contents = targetInv.contents

        player.openInventory(wrapper)

        // TODO: sync changes back into target’s inventory when closing widget
    }

    @Command("absorb")
    fun absorb(actor: BukkitCommandActor) {
        actor.sender().sendRichMessage("<gray>Not yet implemented</gray>")
    }

    @Command("smite")
    fun smite(
        actor: BukkitCommandActor,
        @Named("target") target: Player
    ) {
        val player = actor.sender()
        target.takeIf { it.isOnline }?.let { tgt ->
            try {
                tgt.world.strikeLightning(target.location)
                player.sendRichMessage("<gray>You have smitten <dark_aqua>${target.name}</dark_aqua>!</gray>")
                target.sendRichMessage("<gray>You have been smitten by <b><dark_aqua>a mighty force</b>!</gray>")
            } catch (e: Exception) {
                ChatFormatting.sendError(player, "failed to smite player - ${e.message}")
            }
        } ?: ChatFormatting.sendError(player, "player not found or offline")
    }

    @Command("sudo")
    class SudoCommands(private val plugin: PowerToolsPlugin) {
        @Subcommand("cmd")
        fun sudoCommand(
            actor: BukkitCommandActor,
            @Named("target") target: Player,
            @Named("command") command: String
        ) {
            if (!target.isOnline) {
                ChatFormatting.sendError(actor.sender(), "player not found or offline")
                return
            }
            AsyncCraftr.runEntityTask(plugin, target, Runnable { target.chat("/${command.trim()}") })
        }

        @Subcommand("echo")
        fun sudoChat(
            actor: BukkitCommandActor,
            @Named("target") target: Player,
            @Named("message") message: String
        ) {
            if (!target.isOnline) {
                ChatFormatting.sendError(actor.sender(), "player not found or offline")
                return
            }

            target.chat(message)
        }
    }

    @Command("split", "unenchant")
    fun split(actor: BukkitCommandActor) {
        val player = actor.sender() as Player

        val inventory = player.inventory
        val itemHand = inventory.itemInMainHand

        if (itemHand.type == Material.ENCHANTED_BOOK) {
            ChatFormatting.sendError(player, "the item can't be an enchanted book")
            return
        }

        if (itemHand.type == Material.AIR) {
            ChatFormatting.sendError(player, "you must be holding an item")
            return
        }

        val enchantments = itemHand.enchantments
        if (enchantments.isEmpty()) {
            ChatFormatting.sendError(player, "this item has no enchantments to split")
            return
        }

        val cost = (config.unenchant.basePrice * enchantments.size).roundToInt()
        if (player.totalExperience < cost) {
            ChatFormatting.sendError(player, "you need at least $cost XP to split these enchantments")
            return
        }
        player.giveExp(-cost)

        val meta = itemHand.itemMeta
        enchantments.keys.forEach(meta::removeEnchant)
        itemHand.itemMeta = meta
        itemHand.resetData(DataComponentTypes.REPAIR_COST)

        enchantments.forEach { (ench, level) ->
            val book = ItemStack(Material.ENCHANTED_BOOK)
            val bookMeta = book.itemMeta as EnchantmentStorageMeta
            bookMeta.addStoredEnchant(ench, level, true)
            book.itemMeta = bookMeta

            if (inventory.firstEmpty() == -1) {
                player.world.dropItemNaturally(player.location, book)
            } else {
                inventory.addItem(book)
            }
        }

        player.updateInventory()
        player.sendRichMessage(
            "<gray>Successfully split <enchantments> enchantment(s) for <cost> XP.</gray>",
            Placeholder.component("enchantments", Component.text(enchantments.size.toString(), NamedTextColor.DARK_AQUA)),
            Placeholder.component("cost", Component.text(cost.toString(), NamedTextColor.GREEN))
        )
    }

    @Command("fly")
    fun fly(actor: BukkitCommandActor) {
        val player = actor.sender() as Player
        val canFly = player.allowFlight
        val toggledFly = !canFly

        val statusMessage = if (toggledFly) "enabled" else "disabled"
        val color = if (toggledFly) "green" else "red"
        val playerName = "<dark_aqua>${player.name}</dark_aqua>"

        if (player.gameMode == GameMode.CREATIVE) {
            player.allowFlight = true
            ChatFormatting.sendError(player, "flying is always disabled in creative mode. Keeping fly enabled")
            return
        }
        player.allowFlight = toggledFly
        player.sendRichMessage("<gray>Fly <$color>$statusMessage</$color> for $playerName.</gray>")
    }

    @Command("ue", "unsafe-ench", "uenchant")
    fun enchantUnsafe(
        actor: BukkitCommandActor,
        @SuggestWith(EnchantmentSuggestions::class) @Named("enchantment") enchant: String,
        @Named("level") level: Int
    ) {
        // this should be precomputed somewhere
        val unsafeEnchantsPrefix = ChatFormatting.colorConverter.convertToComponent("&8[&5UE&8]")
        // context
        val registry = RegistryAccess.registryAccess().getRegistry(RegistryKey.ENCHANTMENT)
        val typedKey = TypedKey.create(RegistryKey.ENCHANTMENT, Key.key("minecraft:$enchant"))

        val enchantment = registry.get(typedKey) ?: run {
            ChatFormatting.sendError(actor.sender(), "invalid enchantment, or not found")
            return
        }

        val player = actor.sender() as Player

        if (!config.unsafeEnchants.enabled) {
            ChatFormatting.sendError(player, "unsafe enchantments are disabled in config")
            return
        }

        val item = player.inventory.itemInMainHand
        if (item.type == Material.AIR) {
            ChatFormatting.sendError(player, "you must be holding an item to enchant")
            return
        }

        item.addUnsafeEnchantment(enchantment, level)
        player.updateInventory()

        val displayName = PlainTextComponentSerializer.plainText()
            .serialize(enchantment.displayName(1))

        val finalDisplayName = displayName.trim().split("\\s+".toRegex()).let { words ->
            if (words.isNotEmpty() && romanNumeralRegex.matches(words.last())) {
                words.dropLast(1).joinToString(" ")
            } else {
                displayName
            }
        }

        val finalDisplayNameComponent = Component.text(finalDisplayName, NamedTextColor.DARK_AQUA)
        val levelComponent = Component.text(level.toString(), NamedTextColor.DARK_GREEN)

        player.sendRichMessage(
            "<prefix> <gray>Applied <ench-display-name> at level <level>.",
            Placeholder.component("prefix", unsafeEnchantsPrefix),
            Placeholder.component("ench-display-name", finalDisplayNameComponent),
            Placeholder.component("level", levelComponent)
        )
    }

    @Command("heal", "h")
    fun heal(
        actor: BukkitCommandActor,
        @Optional @Named("player") target: Player?
    ) {
        val healTarget = target ?: (actor.sender() as? Player)
        healTarget?.let { healPlayer(it, actor.sender()) }
    }

    private fun healPlayer(player: Player, sender: CommandSender) {
        player.getAttribute(Attribute.MAX_HEALTH)?.let {
            player.health = it.value
        }
        player.foodLevel = 20
        player.saturation = 20f

        if (config.heal.removeEffects) {
            player.activePotionEffects.forEach { effect -> player.removePotionEffect(effect.type) }
        }

        val healedMessage = "<gray>You have been <dark_aqua>healed</dark_aqua>.</gray>"

        if (player != sender) {
            sender.sendRichMessage(
                "<gray><player> has been healed.</gray>",
                Placeholder.component("player", Component.text(player.name))
            )

            val healMsg = if (config.heal.showWhoHealed) {
                "<gray>You have been healed by <staff>.</gray>"
            } else {
                healedMessage
            }

            player.sendRichMessage(healMsg, Placeholder.component("staff", Component.text(sender.name)))
        } else {
            sender.sendRichMessage(healedMessage)
        }
    }
}

// Configuration classes
data class UnenchantConfig(val basePrice: Double)
data class UnsafeEnchantConfig(val enabled: Boolean)
data class HealConfig(val removeEffects: Boolean, val showWhoHealed: Boolean)
data class TransferConfig(val enabled: Boolean, val servers: List<Map<String, String>>)
data class Configuration(
    val heal: HealConfig,
    val unenchant: UnenchantConfig,
    val unsafeEnchants: UnsafeEnchantConfig,
    val transferConfig: TransferConfig
)
